src/Callable.h

```diff
@@ -0,0 +1,357 @@
+    }
+
+    template<typename... Args>
+    HALIDE_MUST_USE_RESULT int
+    operator()(Args &&...args) const {
+        JITUserContext empty;
+        return call(&empty, std::forward<Args>(args)...);
+    }
+
+    /** This allows us to construct a std::function<> that wraps the Callable.
+     * This is nice in that it is, well, just a std::function, but also in that
+     * since the argument-count-and-type checking are baked into the language,
+     * we can do the relevant checking only once -- when we first create the std::function --
+     * and skip it on all actual *calls* to the function, making it slightly more efficient.
+     * It's also more type-forgiving, in that the usual C++ numeric coercion rules apply here.
+     *
+     * The downside is that there isn't (currently) any way to automatically infer
+     * the static types reliably, since we may be using (e.g.) a Param<void>, where the
+     * type in question isn't available to the C++ compiler. This means that the coder
+     * must supply the correct type signature when calling this function -- but the good news
+     * is that if you get it wrong, this function will fail when you call it. (In other words:
+     * it can't choose the right thing for you, but it can tell you when you do the wrong thing.)
+     *
+     * TODO: it's possible that we could infer the correct signatures in some cases,
+     * and only fail for the ambiguous cases, but that would require a lot more template-fu
+     * here and elsewhere. I think this is good enough for now.
+     */
+    template<typename First, typename... Rest>
+    std::function<int(First, Rest...)>
+    make_std_function() const {
+        if constexpr (std::is_same_v<First, JITUserContext *>) {
+            constexpr auto actual_arg_types = make_fcci_array<First, Rest...>();
+            check_fcci(actual_arg_types.size(), actual_arg_types.data());
+
+            // Capture *this to ensure that the CallableContents stay valid as long as the std::function does
+            return [*this](auto &&first, auto &&...rest) -> int {
+                constexpr size_t count = 1 + sizeof...(rest);
+                ArgvStorage<count> argv(std::forward<First>(first), std::forward<Rest>(rest)...);
+                return call_argv_fast(count, &argv.argv[0]);
+            };
+        } else {
+            // Explicitly prepend JITUserContext* as first actual-arg-type.
+            constexpr auto actual_arg_types = make_fcci_array<JITUserContext *, First, Rest...>();
+            check_fcci(actual_arg_types.size(), actual_arg_types.data());
+
+            // Capture *this to ensure that the CallableContents stay valid as long as the std::function does
+            return [*this](auto &&first, auto &&...rest) -> int {
+                // Explicitly prepend an (empty) JITUserContext to the args.
+                JITUserContext empty;
+                constexpr size_t count = 1 + 1 + sizeof...(rest) + 1;
```

Comment:
What do the three 1s represent? The first is going to be JITUserContext, then there's I guess the argument "first", and the last one is?

Reply:
I think it represents a dumb typo on my part, fixing.